File: CPP/Lee606.cpp
//
// Created by yangf on 2022/3/19.
//
#include <string>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:

    void traverse(TreeNode* node, string& result) {
        if (node == nullptr) {
            return;
        }
        result += to_string(node->val);
        if (node->left == nullptr && node->right == nullptr) {
            return;
        }
        result += '(';
        traverse(node->left, result);
        result += ")(";
        traverse(node->right, result);
        result += ')';
        if (result.substr(result.length() - 2, 2) == "()") {
            result.erase(result.length() - 2);
        }
    }

    string tree2str(TreeNode* root) {
        string result;
        traverse(root, result);
        return result;
    }
};

File: JAVA/Lee606.java
package JAVA;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Lee606 {
    void traverse(TreeNode root, StringBuilder result) {
        if (root == null) {
            return;
        }
        result.append(root.val);
        if (root.left == null && root.right == null) {
            return;
        }
        result.append('(');
        traverse(root.left, result);
        result.append(")(");
        traverse(root.right, result);
        result.append(')');
        if (result.substring(result.length() - 2, result.length()).equals("()")) {
            result.delete(result.length() - 2, result.length());
        }
    }

    public String tree2str(TreeNode root) {
        StringBuilder result = new StringBuilder();
        traverse(root, result);
        return result.toString();
    }
}
